crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+            .qubits
+            .bind(py)
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0 as usize)
+            .collect::<Vec<_>>();
+
+        let mut bc: BitData<Clbit> = BitData::new(py, "clbits".to_string());
+        op1.instruction
+            .clbits
+            .bind(py)
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+        op2.instruction
+            .clbits
+            .bind(py)
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+        let cargs1 = op1
+            .instruction
+            .clbits
+            .bind(py)
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0 as usize)
+            .collect::<Vec<_>>();
+        let cargs2 = op2
+            .instruction
+            .clbits
+            .bind(py)
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0 as usize)
+            .collect::<Vec<_>>();
+
+        Ok(self.commute_inner(
+            &op1.instruction,
+            &qargs1,
+            &cargs1,
+            &op2.instruction,
+            &qargs2,
+            &cargs2,
+            max_num_qubits,
+        ))
+    }
+
+    #[pyo3(signature=(op1, qargs1, cargs1, op2, qargs2, cargs2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute(
+        &mut self,
+        py: Python,
+        op1: &Bound<PyAny>,
+        qargs1: Option<&Bound<PySequence>>,
+        cargs1: Option<&Bound<PySequence>>,
+        op2: &Bound<PyAny>,
+        qargs2: Option<&Bound<PySequence>>,
+        cargs2: Option<&Bound<PySequence>>,
+        max_num_qubits: u32,
+    ) -> PyResult<bool> {
+        let qargs1 =
+            qargs1.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let cargs1 =
+            cargs1.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let qargs2 =
+            qargs2.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let cargs2 =
+            cargs2.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        self.commute_nodes(
+            py,
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op1,
+                    Some(qargs1.into_any()),
+                    Some(cargs1.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
+            },
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op2,
+                    Some(qargs2.into_any()),
+                    Some(cargs2.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
+            },
+            max_num_qubits,
+        )
+    }
+
+    #[pyo3(signature=())]
+    fn num_cached_entries(&self) -> usize {
+        self.current_cache_entries
+    }
```

Comment:
Could this be replaced by `self.cache.len()`?

Reply:
Unfortunately no, I don't think so. The cache can contain another `Dict` that would be filled with commutation resolutions depending on relative qubit placement.